File: python_server/data/storage.py
from datetime import datetime
from typing import List, Dict, Optional, Union, Any
from .models import User, CelestialObject, Observation, MonthlyGuide, TelescopeTip

class MemStorage:
    def __init__(self):
        self.users = {}
        self.celestial_objects = {}
        self.observations = {}
        self.monthly_guides = {}
        self.telescope_tips = {}
        
        # Counters for IDs
        self.user_current_id = 0
        self.object_current_id = 0
        self.observation_current_id = 0
        self.guide_current_id = 0
        self.tip_current_id = 0
    
    # User operations
    def get_user(self, id: int) -> Optional[User]:
        return self.users.get(id)
    
    def get_user_by_username(self, username: str) -> Optional[User]:
        for user in self.users.values():
            if user.username == username:
                return user
        return None
    
    def create_user(self, user_data: Dict[str, Any]) -> User:
        self.user_current_id += 1
        id = self.user_current_id
        
        user = User(
            id=id,
            username=user_data.get('username'),
            email=user_data.get('email'),
            password_hash=user_data.get('passwordHash')
        )
        
        self.users[id] = user
        return user
    
    # Celestial object operations
    def get_celestial_object(self, id: int) -> Optional[CelestialObject]:
        return self.celestial_objects.get(id)
    
    def get_all_celestial_objects(self) -> List[CelestialObject]:
        return list(self.celestial_objects.values())
    
    def get_celestial_objects_by_type(self, type: str) -> List[CelestialObject]:
        return [obj for obj in self.celestial_objects.values() if obj.type == type]
    
    def get_celestial_objects_by_month(self, month: str) -> List[CelestialObject]:
        return [obj for obj in self.celestial_objects.values() if obj.month == month]
    
    def get_celestial_objects_by_hemisphere(self, hemisphere: str) -> List[CelestialObject]:
        return [obj for obj in self.celestial_objects.values() if obj.hemisphere == hemisphere]
    
    def create_celestial_object(self, object_data: Dict[str, Any]) -> CelestialObject:
        self.object_current_id += 1
        id = self.object_current_id
        
        # Convert camelCase keys to snake_case for our Python model
        celestial_object = CelestialObject(
            id=id,
            name=object_data.get('name'),
            type=object_data.get('type'),
            description=object_data.get('description'),
            coordinates=object_data.get('coordinates'),
            month=object_data.get('month'),
            best_viewing_time=object_data.get('bestViewingTime'),
            image_url=object_data.get('imageUrl'),
            visibility_rating=object_data.get('visibilityRating'),
            information=object_data.get('information'),
            constellation=object_data.get('constellation'),
            magnitude=object_data.get('magnitude'),
            hemisphere=object_data.get('hemisphere'),
            recommended_eyepiece=object_data.get('recommendedEyepiece')
        )
        
        self.celestial_objects[id] = celestial_object
        return celestial_object
    
    # Observation operations
    def get_observation(self, id: int) -> Optional[Observation]:
        return self.observations.get(id)
    
    def get_user_observations(self, user_id: int) -> List[Observation]:
        return [obs for obs in self.observations.values() if obs.user_id == user_id]
    
    def create_observation(self, observation_data: Dict[str, Any]) -> Observation:
        self.observation_current_id += 1
        id = self.observation_current_id
        date_added = datetime.now()
        
        observation = Observation(
            id=id,
            user_id=observation_data.get('userId'),
            object_id=observation_data.get('objectId'),
            date_added=date_added,
            is_observed=observation_data.get('isObserved', False),
            observation_notes=observation_data.get('observationNotes'),
            planned_date=observation_data.get('plannedDate')
        )
        
        self.observations[id] = observation
        return observation
    
    def update_observation(self, id: int, update_data: Dict[str, Any]) -> Optional[Observation]:
        observation = self.observations.get(id)
        if not observation:
            return None
        
        # Update the fields
        if 'isObserved' in update_data:
            observation.is_observed = update_data['isObserved']
        
        if 'observationNotes' in update_data:
            observation.observation_notes = update_data['observationNotes']
        
        if 'plannedDate' in update_data:
            observation.planned_date = update_data['plannedDate']
        
        return observation
    
    def delete_observation(self, id: int) -> bool:
        if id in self.observations:
            del self.observations[id]
            return True
        return False
    
    # Monthly guide operations
    def get_monthly_guide(self, id: int) -> Optional[MonthlyGuide]:
        return self.monthly_guides.get(id)
    
    def get_current_monthly_guide(self, hemisphere: str) -> Optional[MonthlyGuide]:
        current_month = datetime.now().strftime('%B')
        current_year = datetime.now().year
        
        for guide in self.monthly_guides.values():
            if (guide.month == current_month and 
                guide.year == current_year and 
                (guide.hemisphere == hemisphere or guide.hemisphere == 'both')):
                return guide
        
        return None
    
    def get_all_monthly_guides(self) -> List[MonthlyGuide]:
        return list(self.monthly_guides.values())
    
    def create_monthly_guide(self, guide_data: Dict[str, Any]) -> MonthlyGuide:
        self.guide_current_id += 1
        id = self.guide_current_id
        
        guide = MonthlyGuide(
            id=id,
            month=guide_data.get('month'),
            year=guide_data.get('year'),
            headline=guide_data.get('headline'),
            content=guide_data.get('content'),
            hemisphere=guide_data.get('hemisphere'),
            featured_objects=guide_data.get('featuredObjects', [])
        )
        
        self.monthly_guides[id] = guide
        return guide
    
    # Telescope tip operations
    def get_telescope_tip(self, id: int) -> Optional[TelescopeTip]:
        return self.telescope_tips.get(id)
    
    def get_all_telescope_tips(self) -> List[TelescopeTip]:
        return list(self.telescope_tips.values())
    
    def get_telescope_tips_by_category(self, category: str) -> List[TelescopeTip]:
        return [tip for tip in self.telescope_tips.values() if tip.category == category]
    
    def create_telescope_tip(self, tip_data: Dict[str, Any]) -> TelescopeTip:
        self.tip_current_id += 1
        id = self.tip_current_id
        
        tip = TelescopeTip(
            id=id,
            title=tip_data.get('title'),
            content=tip_data.get('content'),
            category=tip_data.get('category'),
            image_url=tip_data.get('imageUrl')
        )
        
        self.telescope_tips[id] = tip
        return tip

# Create a singleton instance
storage = MemStorage()

File: server/storage.ts
import {
  User, InsertUser,
  CelestialObject, InsertCelestialObject,
  Observation, InsertObservation,
  MonthlyGuide, InsertMonthlyGuide,
  GuideObject, InsertGuideObject,
  TelescopeTip, InsertTelescopeTip,
  users, celestialObjects, observations,
  monthlyGuides, guideObjects, telescopeTips, apodCache
} from "@shared/schema";
import { db } from "./db";
import { eq, and, or, inArray } from "drizzle-orm";

// Extend the storage interface with all required CRUD operations
export interface IStorage {
  // User operations
  getUser(id: number): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User | undefined>;
  createUser(user: InsertUser): Promise<User>;
  
  // Celestial objects operations (static catalog)
  getCelestialObject(id: number): Promise<CelestialObject | undefined>;
  getCelestialObjectByName(name: string): Promise<CelestialObject | undefined>;
  getAllCelestialObjects(): Promise<CelestialObject[]>;
  getCelestialObjectsByType(type: string): Promise<CelestialObject[]>;
  getCelestialObjectsByIds(ids: number[]): Promise<CelestialObject[]>;
  createCelestialObject(object: InsertCelestialObject): Promise<CelestialObject>;
  updateCelestialObject(id: number, object: Partial<CelestialObject>): Promise<CelestialObject | undefined>;
  deleteCelestialObject(id: number): Promise<boolean>;
  
  // Observation operations
  getObservation(id: number): Promise<Observation | undefined>;
  getUserObservations(userId: number): Promise<Observation[]>;
  createObservation(observation: InsertObservation): Promise<Observation>;
  updateObservation(id: number, observation: Partial<Observation>): Promise<Observation | undefined>;
  deleteObservation(id: number): Promise<boolean>;
  
  // Monthly guides operations
  getMonthlyGuide(id: number): Promise<MonthlyGuide | undefined>;
  getMonthlyGuideByMonthYear(month: string, year: number, hemisphere: string): Promise<MonthlyGuide | undefined>;
  getCurrentMonthlyGuide(hemisphere: string): Promise<MonthlyGuide | undefined>;
  getAllMonthlyGuides(): Promise<MonthlyGuide[]>;
  createMonthlyGuide(guide: InsertMonthlyGuide): Promise<MonthlyGuide>;
  updateMonthlyGuide(id: number, guide: Partial<MonthlyGuide>): Promise<MonthlyGuide | undefined>;
  deleteMonthlyGuide(id: number): Promise<boolean>;

  // Guide objects operations (links objects to guides)
  getGuideObjects(guideId: number): Promise<GuideObject[]>;
  getGuideObjectsWithDetails(guideId: number): Promise<(GuideObject & { object: CelestialObject })[]>;
  createGuideObject(guideObject: InsertGuideObject): Promise<GuideObject>;
  deleteGuideObjectsByGuide(guideId: number): Promise<boolean>;
  
  // Telescope tips operations
  getTelescopeTip(id: number): Promise<TelescopeTip | undefined>;
  getAllTelescopeTips(): Promise<TelescopeTip[]>;
  getTelescopeTipsByCategory(category: string): Promise<TelescopeTip[]>;
  createTelescopeTip(tip: InsertTelescopeTip): Promise<TelescopeTip>;
}

export class DatabaseStorage implements IStorage {
  async getUser(id: number): Promise<User | undefined> {
    const [user] = await db.select().from(users).where(eq(users.id, id));
    return user || undefined;
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    const [user] = await db.select().from(users).where(eq(users.username, username));
    return user || undefined;
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const [user] = await db
      .insert(users)
      .values(insertUser)
      .returning();
    return user;
  }

  async getCelestialObject(id: number): Promise<CelestialObject | undefined> {
    const [object] = await db.select().from(celestialObjects).where(eq(celestialObjects.id, id));
    return object || undefined;
  }
  
  async getCelestialObjectByName(name: string): Promise<CelestialObject | undefined> {
    const [object] = await db.select().from(celestialObjects).where(eq(celestialObjects.name, name));
    return object || undefined;
  }

  async getAllCelestialObjects(): Promise<CelestialObject[]> {
    return await db.select().from(celestialObjects);
  }

  async getCelestialObjectsByType(type: string): Promise<CelestialObject[]> {
    return await db.select().from(celestialObjects).where(eq(celestialObjects.type, type));
  }

  async getCelestialObjectsByIds(ids: number[]): Promise<CelestialObject[]> {
    if (ids.length === 0) return [];
    return await db.select().from(celestialObjects).where(inArray(celestialObjects.id, ids));
  }

  async createCelestialObject(insertObject: InsertCelestialObject): Promise<CelestialObject> {
    const [object] = await db
      .insert(celestialObjects)
      .values(insertObject)
      .returning();
    return object;
  }

  async updateCelestialObject(id: number, update: Partial<CelestialObject>): Promise<CelestialObject | undefined> {
    const [updatedObject] = await db
      .update(celestialObjects)
      .set(update)
      .where(eq(celestialObjects.id, id))
      .returning();
    return updatedObject || undefined;
  }

  async deleteCelestialObject(id: number): Promise<boolean> {
    const [deleted] = await db
      .delete(celestialObjects)
      .where(eq(celestialObjects.id, id))
      .returning();
    return !!deleted;
  }

  async getObservation(id: number): Promise<Observation | undefined> {
    const [observation] = await db.select().from(observations).where(eq(observations.id, id));
    return observation || undefined;
  }

  async getUserObservations(userId: number): Promise<Observation[]> {
    return await db.select().from(observations).where(eq(observations.userId, userId));
  }

  async createObservation(insertObservation: InsertObservation): Promise<Observation> {
    const [observation] = await db
      .insert(observations)
      .values(insertObservation)
      .returning();
    return observation;
  }

  async updateObservation(id: number, update: Partial<Observation>): Promise<Observation | undefined> {
    const [updatedObservation] = await db
      .update(observations)
      .set(update)
      .where(eq(observations.id, id))
      .returning();
    return updatedObservation || undefined;
  }

  async deleteObservation(id: number): Promise<boolean> {
    const [deleted] = await db
      .delete(observations)
      .where(eq(observations.id, id))
      .returning();
    return !!deleted;
  }

  async getMonthlyGuide(id: number): Promise<MonthlyGuide | undefined> {
    const [guide] = await db.select().from(monthlyGuides).where(eq(monthlyGuides.id, id));
    return guide || undefined;
  }

  async getMonthlyGuideByMonthYear(month: string, year: number, hemisphere: string): Promise<MonthlyGuide | undefined> {
    const [guide] = await db.select().from(monthlyGuides).where(
      and(
        eq(monthlyGuides.month, month),
        eq(monthlyGuides.year, year),
        or(
          eq(monthlyGuides.hemisphere, hemisphere),
          eq(monthlyGuides.hemisphere, 'Both')
        )
      )
    );
    return guide || undefined;
  }

  async getCurrentMonthlyGuide(hemisphere: string): Promise<MonthlyGuide | undefined> {
    const currentMonth = new Date().toLocaleString('default', { month: 'long' });
    const currentYear = new Date().getFullYear();
    return this.getMonthlyGuideByMonthYear(currentMonth, currentYear, hemisphere);
  }

  async getAllMonthlyGuides(): Promise<MonthlyGuide[]> {
    return await db.select().from(monthlyGuides);
  }

  async createMonthlyGuide(insertGuide: InsertMonthlyGuide): Promise<MonthlyGuide> {
    const [guide] = await db
      .insert(monthlyGuides)
      .values(insertGuide)
      .returning();
    return guide;
  }

  async updateMonthlyGuide(id: number, update: Partial<MonthlyGuide>): Promise<MonthlyGuide | undefined> {
    const [updatedGuide] = await db
      .update(monthlyGuides)
      .set(update)
      .where(eq(monthlyGuides.id, id))
      .returning();
    return updatedGuide || undefined;
  }

  async deleteMonthlyGuide(id: number): Promise<boolean> {
    // First delete all guide objects for this guide
    await db.delete(guideObjects).where(eq(guideObjects.guideId, id));
    // Then delete the guide itself
    const [deleted] = await db
      .delete(monthlyGuides)
      .where(eq(monthlyGuides.id, id))
      .returning();
    return !!deleted;
  }

  // Guide objects operations
  async getGuideObjects(guideId: number): Promise<GuideObject[]> {
    return await db.select().from(guideObjects)
      .where(eq(guideObjects.guideId, guideId))
      .orderBy(guideObjects.sortOrder);
  }

  async getGuideObjectsWithDetails(guideId: number): Promise<(GuideObject & { object: CelestialObject })[]> {
    const gos = await this.getGuideObjects(guideId);
    const objectIds = gos.map(go => go.objectId);
    const objects = await this.getCelestialObjectsByIds(objectIds);
    const objectMap = new Map(objects.map(o => [o.id, o]));

    return gos.map(go => ({
      ...go,
      object: objectMap.get(go.objectId)!
    })).filter(go => go.object !== undefined);
  }

  async createGuideObject(insertGuideObject: InsertGuideObject): Promise<GuideObject> {
    const [guideObject] = await db
      .insert(guideObjects)
      .values(insertGuideObject)
      .returning();
    return guideObject;
  }

  async deleteGuideObjectsByGuide(guideId: number): Promise<boolean> {
    await db.delete(guideObjects).where(eq(guideObjects.guideId, guideId));
    return true;
  }

  async getTelescopeTip(id: number): Promise<TelescopeTip | undefined> {
    const [tip] = await db.select().from(telescopeTips).where(eq(telescopeTips.id, id));
    return tip || undefined;
  }

  async getAllTelescopeTips(): Promise<TelescopeTip[]> {
    return await db.select().from(telescopeTips);
  }

  async getTelescopeTipsByCategory(category: string): Promise<TelescopeTip[]> {
    return await db.select().from(telescopeTips).where(eq(telescopeTips.category, category));
  }

  async createTelescopeTip(insertTip: InsertTelescopeTip): Promise<TelescopeTip> {
    const [tip] = await db
      .insert(telescopeTips)
      .values(insertTip)
      .returning();
    return tip;
  }
}

// Use DatabaseStorage instead of MemStorage for persistent storage
export const storage = new DatabaseStorage();